rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,386 @@
+
+However, unlike Per-session MFA, which scopes certificates to a specific node connection, we will need to scope the certificates to the entire API flow for a `tsh` command. For example, `tsh ssh` needs to:
+
+* Connect to Proxy
+* List nodes
+* View cluster details (Auth preference, etc.)
+* Connect to Node
+
+This means that for any given `tsh` command, the Teleport Auth server will provide different privileges to the certificates. Since each possible command will need custom logic to determine the lowest privileges necessary, we will only support a select few commands necessary for basic `tsh` workflows:
+
+* `tsh ls`
+* `tsh ssh`
+* `tsh scp`
+
+The awarded certificates should not provide access to roles usually awarded to the user, as these roles may provide more privileges than required for the requested command. However, some role provided fields should be included, such as `role.options.forward_agent` and `role.logins`. Additionally, the user should be notified of what command was requested and what permissions will be awarded to the headless session once approved.
+
+Example:
+
+```
+Incoming headless request for command "tsh ssh server01", to connect to node <server01_uuid>. Upon approval, the request will be awarded the following permissions:
+
+allow:
+  options:
+    max_session_ttl: 8h0m0s
+    forward_agent: true
+    port_forwarding: true
+    require_session_mfa: true
+  logins: ["ubuntu-user"]
+  resource_ids: ["<server01_uuid>"]
+  rules:
+    resources: ["nodes"]
+    verbs: ["read", "list"]
+```
+
+**Important note**: The server cannot guarantee that the client will actually use the resulting certificates for the `tsh` command advertised by the client. This means that a modified client could claim to need certificates for `tsh ls`, but then actually use the certificates for `tsh ssh`. In order to prevent potential permission workarounds, we will need to carefully consider how each supported command's awarded certificates may interact with other Teleport clients. For this reason, we may add more commands as they are requested, but each command may have significant security implications to consider. Each PR adding one or more new commands should include a security description and get a security review from the Teleport security team. This includes the commands listed above, as the most limited necessary permissions necessary to perform each command is not yet clear.
+
+#### Caveat
+
+Due to the implementation complexity of limiting certificates permissions by requested command, it will not be included in the feature's preview release.
+
+Additionally, some details, like how the server will encode the permissions into the certificate, are currently missing. This will be addressed in a follow up PR before the official release of this feature, accompanied by an RFD update.
```

Comment:
This is the opposite of why we do RFDs. Prototyping is fine and you should do it, but the design should come before the end product.

Reply:
Agreed, I would prefer not to do it this way. For background, we are on a tighter timeline than usual for this feature, so some shortcuts are required.

Reply:
It sounds like you're still not sure how some parts of this are going to work, and that's why the detail is currently missing from the RFD.

You have to figure out these missing pieces either way, so I fail to see how discovering the unknowns now and updating the RFD is any faster then discovering them later and updating the RFD after. From my point of view, it seems both options are roughly the same amount of work but one of them allows the team to help you evaluate the design decisions early and one of them doesn't. The former is much less risky, no?

Reply:
We will keep thinking about the design of this particular requirement (limiting certs scope) and definitely solidify its design before starting its implementation. It is not in scope for the initial release so for now @Joerger let's please push it out of scope of the RFD so we don't block the implementation. cc @jakule

I agree with you Alan that this can be built as a separate subsystem. The more I think about it the more it seems to me that supporting something like this will require us to introduce some concept of "permission boundary" (or similar) in Teleport RBAC which will likely require a whole other research and a separate RFD.